src/flavors/list.rs

```diff
@@ -55,34 +39,74 @@ impl Default for ListToken {
     }
 }
 
+/// A slot in a block.
+struct Slot<T> {
+    /// The message.
+    msg: UnsafeCell<ManuallyDrop<T>>,
+
+    /// Equals `true` if the message is ready for reading.
+    ready: AtomicBool,
+}
+
 /// A block in a linked list.
 ///
-/// Each block in the list can hold up to `slots.len()` messages. `slots.len()` will double for
-/// each consecutive block - starting at 1 and growing no larger than `BLOCK_CAP`.
+/// Each block in the list can hold up to `cap` slots. `cap` will double for each consecutive block
+/// - starting at 1 and growing no larger than `BLOCK_CAP`.
+#[repr(C)]
 struct Block<T> {
     /// The start index of this block.
     ///
-    /// Slots in this block have indices in `start_index .. start_index + slots.len()`.
+    /// Slots in this block have indices in `start_index .. start_index + cap`.
     start_index: usize,
 
+    /// The number of slots this block can hold.
+    cap: usize,
+
     /// The next block in the linked list.
     next: Atomic<Block<T>>,
 
     /// Slots for messages.
-    slots: Box<[Slot<T>]>,
+    slots: [Slot<T>; 0],
 }
 
 impl<T> Block<T> {
     /// Creates an empty block that starts at `start_index`.
-    fn new(start_index: usize, slot_count: usize) -> Block<T> {
-        Block {
-            start_index,
-            slots: (0..slot_count)
-                .map(|_| Slot::empty())
-                .collect::<Vec<_>>()
-                .into_boxed_slice(),
-            next: Atomic::null(),
-        }
+    fn new<'a>(start_index: usize, cap: usize) -> Shared<'a, Block<T>> {
+        let block = unsafe {
+            let ptr = Self::alloc(cap);
+
+            // Write the `start_index` and `cap`. Initialize `next` to null and zero out `slots`.
+            ptr::write(&mut (*ptr).start_index, start_index);
+            ptr::write(&mut (*ptr).cap, cap);
+            ptr::write(&mut (*ptr).next, Atomic::null());
+            ptr::write_bytes((*ptr).slots.as_mut_ptr(), 0, cap);
+
+            ptr
+        };
+
+        Shared::from(block as *const _)
+    }
+
+    /// Allocate memory for a block with space to hold `cap` slots.
+    unsafe fn alloc(cap: usize) -> *mut Self {
+        let layout = Self::create_layout(cap);
+        // TODO(kleimkuhler): Use `GlobalAlloc::alloc` once `Global` becomes stable
+        alloc::alloc(layout) as *mut Self
+    }
+
+    /// Deallocate memory that a block with `cap` slots holds.
```

Comment:
Nit: There is no `cap` argument here. :)

Reply:
Ah yes was being a little too specific there; it's gone!